### PR TITLE
Porting txstate

### LINKED ISSUE #53: Suggestion: Less awkward way to pass context and event types
> useStateMachine is a curried function (Yummm tasty!) because TypeScript doesn't yet support partial generics type inference. This workaround allows TypeScript developers to provide a custom type for the context while still having TypeScript infer all the types used in the configuration (Like the state & transitions names, etc...).

We no longer have this limitation with #36. We can have something like the following for context...

```ts
useStateMachine({
  initial: "idle",
  context: { foo: 1 } // I prefer "initialContext" but whatever xD
  ...
})
```

And something like [this (what xstate v5 does)](https://github.com/davidkpiano/xstate/pull/1977), for events and context...

```ts
useStateMachine({
  schema: {
    context: createSchema<{ foo: number, bar?: number }>(),
    event: createSchema<
      | { type: "X", foo: number }  
      | { type: "Y", bar: number }
    >(),
    // I prefer "event" instead of "events" because the type is for event and not "events"
    // (unions need to be named singular even though they seem plural)
    // though "events" works too if that's more friendly
  },
  initial: "idle",
  context: { foo: 1 }
  ...
})
```

### PR DESCRIPTION
**Pull request by devanshj**
Superseded by #56 

## Todos
- [x] Add tests for top level `on`
- [x] Fix the `_` bug - right now we need to add `_: null` is a state node only has a effect, I think the problem is at TS's end though will have to find a workaround. - update: made it `on: null` ~still haven't investigated yet to see if we can fix it completely~ we can't fix it, so pending is to add a custom error to tell user to add `on: null`
- [x] Integrate with [twoslash](https://github.com/microsoft/TypeScript-Website/tree/v2/packages/ts-twoslasher) for `expectCompletionsToBe`, `expectQueryTextToInclude` ie automated testing with language server
- ~(note to self) `type _Effect<D, P> = Effect<X, Y, X>`, etc~ nope, doesn't make intellisense queries more readable
- ~add jsdoc~ apparently there is no effect in adding jsdocs to the types, they get lost and the user can't see them
- [x] Typestated `state`
- [x] schema and new schema closes #53 

## Original Description

Hi Cassio! Here's a little rough PR for porting [txstate](https://github.com/devanshj/txstate) for better types.

#### Way better error messages
<details><summary><strong>Scenario 1</strong></summary>

<br>Before - 
image
image

After -
image
image
</details>
<details><summary><strong>Scenario 2</strong></summary>

<br>Before - 
image
image

After -
image
image
</details>

And many more scenarios that I don't have handy...

#### Typed guards hence inferring context typestates

<details><summary><strong>Example</strong></summary>

```ts
// example picked from xstate docs
let [flightMachine] = useStateMachine<FlightMachineContext>({ trip: 'oneWay' })({
  initial: 'editing',
  states: {
    editing: {
      on: {
        SUBMIT: {
          target: 'submitted',
          guard: (context): context is FlightMachineSubmittedContext => {
            if (context.trip === 'oneWay') {
              return !!context.startDate;
            } else {
              return (
                !!context.startDate &&
                !!context.returnDate &&
                context.returnDate > context.startDate
              );
            }
          }
        }
      }
    },
    submitted: {}
  }
})

type FlightMachineEditingContext =
  { trip: 'oneWay' | 'roundTrip'
  , startDate?: Date
  , returnDate?: Date
  }
type FlightMachineSubmittedContext =
  | { trip: 'oneWay'
    , startDate: Date
    }
  | { trip: 'roundTrip'
    , startDate: Date
    , returnDate: Date
    }

type FlightMachineContext =
  | FlightMachineEditingContext
  | FlightMachineSubmittedContext

expectType<FlightMachineContext>(flightMachine.context);
if (flightMachine.value === 'submitted') {
  expectType<FlightMachineSubmittedContext>(flightMachine.context);
  
  if (flightMachine.context.trip === 'oneWay') {
    expectError(flightMachine.context.returnDate)
  }
}
```
</details>

### Other notes
- You can refer to this [document](https://github.com/devanshj/txstate/blob/master/docs/the-foundation.md) to understand how the code works, I hope it helps xD.
- Maybe more tests?
- The PR is purposely opinionated at styles, naming, etc. Everything is negotiable.
- No pressure to get this in, just wanted to know if you're interested.

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -143,9 +144,6 @@ function useStateMachineImpl<Context>(context: Context): UseStateMachineWithCont
...
 }
 
-export default function useStateMachine(): UseStateMachineWithContext<undefined>;
-export default function useStateMachine<Context>(context: Context): UseStateMachineWithContext<Context>;
-export default function useStateMachine<Context>(context?: Context): UseStateMachineWithContext<Context | undefined>;
-export default function useStateMachine<Context>(context?: Context): UseStateMachineWithContext<Context | undefined> {
+export default (function useStateMachine<Context>(context?: Context): UseStateMachineWithContext<Context | undefined> {
   return useStateMachineImpl(context);
-}
+}) as any as UseStateMachine
```
Comment (LXSMNSYC) [inline src/index.tsx:149]: May I ask why is this preferred over function overloads? Or does this work similar with function overloads?

**Comment by cassiozen** [comment]
Hi @devanshj - I'm still trying to dive into this, I have a couple of questions if you don't mind:

1 - You left pretty much all the types intact in `index.tsx` - but I'm assuming these aren't needed anymore?
2 - You have some helper types (great stuff, BTW) under the namespaces O, L, F and A. Why are they under so many namespaces?
3 - I'll be honest with you - I really want to use this, but I don't feel like I can contribute right now to your project. So, my question is - can you keep helping with changes until we push version 1.0? Here are the expected changes:
  -  New event system with payload (Already merged)
  - Top-level events/transitions (#41) 
  - New effects parameter signature with access to context (#33)
  - Nested/hierarchical states (#21)
  If you can and are interested we can coordinate changes on how to change the implementation and types for these changes, and by then I think I'll be more prepared to make changes into `txstate` directly.

**Comment by devanshj** [comment]
@cassiozen Before I reply to you I think there's a big misunderstanding xD You realise this PR has nothing to do with txstate per se, I'm calling it "Porting txstate" because I'm using the techniques used in txstate. This PR and txstate are TOTALLY independent. I linked that document because it'll help you understand this PR somewhat.

So I'm not asking you to contribute to txstate, this PR is to be treated independent as if txstate does not exist.

Did you already know all this or there was some misunderstanding here? xD

**Comment by devanshj** [comment]
To answer your questions -

> You left pretty much all the types intact in index.tsx - but I'm assuming these aren't needed anymore?

The thing is I've kept the user facing types and types for implementation different. The reason is you can't use the one's I've added (I'm calling it user facing) to type implementation details (Let me know if you want me to elaborate on this). So to answer your question we'll be needing both of them.

> You have some helper types (great stuff, BTW) under the namespaces O, L, F and A. Why are they under so many namespaces?

It's inspired from [ts-toolbelt](https://github.com/millsp/ts-toolbelt). The idea is all types under `O` namespace serve as operations on object, `L` on arrays/tuples, `F` in functions and `A` on all types. It's just a matter of convention and preference.

> can you keep helping with changes until we push version 1.0? 

More or less yes, I can't give a word because I haven't checked the issues yet but it's mostly a yes from me. The worst case scenario is we'd have to compromise on types in scenarios where it's not possible to type them. I'll give you a concrete answer once I check issues you've linked.

**Comment by cassiozen** [comment]
Yes, I understand that, my writing was confusing. Whenever I say "contributing to txstate" I mean "contributing to 'types.ts' inside this project".

**Comment by devanshj** [comment]
> Yes, I understand that, my writing was confusing. Whenever I say "contributing to txstate" I mean "contributing to 'types.ts' inside this project".

Ah gotcha, I too thought maybe you meant this but wasn't sure xD

**Comment by devanshj** [comment]
> I'll give you a concrete answer once I check issues you've linked.

I went through all four issues seems doable to me. So it's a yes I'm keep updating this PR (or opening new one's if we're merging feature by feature) till 1.0 (or even forever if time permits :P)

Just there's on issue that the types assume that users are writing the machine definition (config) _in place as literal_ so that the type inferred of the definition is [unit](https://en.wikipedia.org/wiki/Unit_type), as in anything like the following might create an issue...

```ts
let m = { ... } // the definition here
useStateMachine()(m)
// won't work because the types of `m` are already inferred
// so we can't have contextual inference which might cause problems
// because things in `m` would be inferred as `string` (which makes the type non-unit)
// instead of literals like "red"` or whatever

useStateMachine()({
  initial: Math.random() > 0.5 ? "a" : "b", 
  states: { a: {}, b: {} }
})
// might be a problem because the type of machine definition inferred
// is not unit it's a union of two types one with "a" and other with "b".
```

I haven't investigated how to go about cases like this but this could be a potential blocker for this PR, imo the pros are way more than this one con, but you're the boss :P

**Comment by devanshj** [comment]
> @cassiozen Before I reply to you I think there's a big misunderstanding xD You realise this PR has nothing to do with txstate per se, I'm calling it "Porting txstate" because I'm using the techniques used in txstate. This PR and txstate are TOTALLY independent. I linked that document because it'll help you understand this PR somewhat.

> So I'm not asking you to contribute to txstate, this PR is to be treated independent as if txstate does not exist.

> Did you already know all this or there was some misunderstanding here? xD

Also @cassiozen I realized this sounded a little rude when I reread later, I wasn't in the best headspace when I replied, my sincere apologies 🙏😅

**Comment by cassiozen** [comment]
No, you didn't sound rude at all!

Ok, so we have everything merged on the `main` branch for the 1.0 release, whenever you have time to update the types, we can release it as is and leave hierarchical state machines for next on queue. 

A few comments on your responses:

User facing types: Ok, perfect. It worries me a little bit that future developments (like the hierarchical state machines) will have to be done on both sides. Maybe there is a way for me to reuse some of the ported txstate types as user-facing?

Namespaces: Got it, makes sense. In the current types, I moved from single-letter to fully named generics - Do you have strong feelings about this? I could refactor the types.ts names, keeping the same structure.

**Comment by devanshj** [comment]
Cool, going to work on types now hopefully I'll update the branch soon.

> Maybe there is a way for me to reuse some of the ported txstate types as user-facing?

In theory, yes. Thought it'll be extremely verbose and perhaps would do more harm than good (like you'd need more assertions in your implementations.

By verbose I mean it'll look like...
```ts
const someImplementationDetail = <N, C>(node: MachineDefinition.StateNode<N, C, []>) => {}
```
Or in future versions like...
```ts
const someImplementationDetail = <N, C>(node: MachineDefinition.StateNode<N, C, MachineDefinition.Precomputed<N>, []>) => {}
```
Meaning there would be no handy generic-free type that you'd be able to assign to a variable and call it a day.

Instead what I would suggest is dumb down the implementation types. Make them generic-free and just "runtime equivalent" like so...

```ts
interface StateNode {
  initial: string,
  states: Record<string, StateNode>
}
```

Because there isn't a point of having sophisticated types for implementation, in my opinion. If you could manage to have implementation types as good as user facing, well and good, though I personally wouldn't care so much.

> Do you have strong feelings about this?

I kinda have, though we're more or less on the same page, I too am for fully named generic but in few cases I prefer them to be single letter. For example right now more or less all types start with `D, C, S` or at least `D, C`. It wouldn't make sense to write `Definition, Context, Self` all the time. Another instance is for the types inside `O`, `L`, etc namespaces there too it's obvious what single generics types mean. Otherwise I too am for fully named generics like you'll see `States`, `Gaurd`, etc in the `types.ts`. Though in some places they are a little ambiguous like `Ts`, `S`, etc. and I'll refactor them myself.

**Comment by devanshj** [comment]
Hi @cassiozen here's some wip, I'll keep updating the original description and ping you whenever I do it (not sure if github notifies for new commits in a PR), still 15% work more to go

**Comment by cassiozen** [comment]
Yes, I'm super curious to check it out. I'm a little blocked because of worked but will check as soon as possible. 

As for TSD, that's totally fine. I wanted a way to test the types and make sure we make no regressions during new development - but if you can test them using something else that's fine.

**Comment by devanshj** [comment]
Yeah no worries take your time it's unpaid work after all ;)

**Comment by cassiozen** [comment]
@devanshj - I just went through the tests and through the initial work on types.ts. So far everything looks good. I love how you use string literals instead of `never` to represent errors in the type level. 

I plan to start tackling the code to use primitive types in the end of next week.

**Comment by devanshj** [comment]
Nice, I too will complete the rest todos soonish!

> I love how you use string literals instead of never to represent errors in the type level.

Haha ikr thanks!

Also if I can't find a workaround for the "_" thing would it be a blocker? Or it's fine to expect users to write `_: null`?

**Comment by cassiozen** [comment]
Wait, I don't understand. Is that for states with no transitions?

**Comment by devanshj** [comment]
Yep, if `effect` is the only property (ie no `on`) you'd have to add a phantom `_: null` to make it work 😬

**Comment by cassiozen** [comment]
I kind of understand the issue, but it's definitely beyond my TS knowledge. Question, instead of the phantom `_`, does it work if the user is required to provide a null `on` key, or a final state like `type: 'final'`?

**Comment by cassiozen** [comment]
"type: final" would have different implications, so I Think the least bad would be an "on: null".

**Comment by devanshj** [comment]
Also in case you suggested `on: null` thinking it _has_ to be `null` then no, `on: {}` works too. Should we keep that? Or both?

**Comment by cassiozen** [comment]
I think we could keep both.

**Comment by devanshj** [comment]
I think this is ready for final review and merging!

Also I kept it as `on: {}` and not `on: null`, because allowing the latter seems to be that we're changing runtime semantics unnecessarily for type-land issues. And I also added a custom error that instructs the user to add `on: {}` when required you can see the tests for more.

**Comment by cassiozen** [comment]
That is awesome, @devanshj - you're the best!

So my plan is to release this as a beta for at least a week, so you'll have time to write the nice document!

I also want to do a clean-up pass on `index.tsx` to remove what now are unnecessary type definitions and either use primitives or try to use some TXState types passing some `any`s to fill in the generics.

**Comment by devanshj** [comment]
> That is awesome, @devanshj - you're the best!

Haha thanks a lot, my pleasure! And thanks to you too for being so welcoming!

> So my plan is to release this as a beta for at least a week, so you'll have time to write the nice document!

Great, I think a week would be sufficient time.

> I also want to do a clean-up pass on index.tsx to remove what now are unnecessary type definitions and either use primitives or try to use some TXState types passing some anys to fill in the generics.

Yep, feel free to ask questions if you're trying to reuse the types.

**Comment by cassiozen** [comment]
Wait, it seems to be something wrong: I can't find the `createSchema` function...

**Comment by devanshj** [comment]
That's right I replaced it with a more minimal `t` function because the user is going to write them many times so it doesn't make sense to have such long name imo.

File: tsconfig.json [hunk 2]
```diff
@@ -1,33 +1,22 @@
...
     "lib": ["dom", "esnext"],
     "importHelpers": true,
-    // output .d.ts declaration files for consumers
     "declaration": true,
-    // output .js.map sourcemap files for consumers
     "sourceMap": true,
-    // match output dir to input dir. e.g. dist/index instead of dist/src/index
-    "rootDir": "./src",
-    // stricter type-checking for stronger correctness. Recommended by TS
+    // "rootDir": "./src",
```
Comment (devanshj) [inline tsconfig.json:9]: I think removing `"rootDir": "./src"` will affect the build, but I want to type-check "types" folder too. How to go about this?

**Comment by devanshj** [comment]
Also you haven't addressed [this](https://github.com/cassiozen/useStateMachine/pull/36#issuecomment-862297909). And I have added some tasks here too. (I'll probably make type changes here)

File: twoslash-tester/generate.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
+// @ts-check
+
+// @ts-ignore
```
Comment (devanshj) [inline twoslash-tester/generate.js:3]: Note to self
```suggestion
```

File: test/types.twoslash-test.ts [hunk 4]
```diff
@@ -0,0 +1,1093 @@
+/* eslint-disable react-hooks/rules-of-hooks */
+import { A, LS, UseStateMachine } from "../src/types";
+
+const useStateMachine = (() => []) as any as UseStateMachine;
+const t = <T>() => null as any as T
+
+const query = () => (global as any).twoSlashQueries.shift()
```
Comment (devanshj) [inline test/types.twoslash-test.ts:7]: Note to self -
```suggestion
const query = () => (global as any).twoSlashQueries.shift() as { completions: string[], text: string }
```

**Comment by cassiozen** [comment]
Should we close this PR in favor of #56 ?

**Comment by devanshj** [comment]
Yep we should, closing.

**Comment by devanshj** [comment]
[Tho this is left here](#issuecomment-886310932)

**Comment by cassiozen** [comment]
Let me address those separately:

```ts
useStateMachine()({
  initial: Math.random() > 0.5 ? "a" : "b", 
  states: { a: {}, b: {} }
})
// might be a problem because the type of machine definition inferred
// is not unit it's a union of two types one with "a" and other with "b".
```

Absolutely OK.


```ts
let m = { ... } // the definition here
useStateMachine()(m)
// won't work because the types of `m` are already inferred
// so we can't have contextual inference which might cause problems
// because things in `m` would be inferred as `string` (which makes the type non-unit)
// instead of literals like "red"` or whatever
```

I'm ok with this one as well, though we might want to explicitly document this somewhere. 
Question: Would it work if the user defined the configuration externally but used `as const` everywhere?

**Comment by devanshj** [comment]
> though we might want to explicitly document this somewhere.

We'd document both, I'll add it to the "downsides" section of my release article.

> Would it work if the user defined the configuration externally but used as const everywhere?

Yep that'd work, though consequently the errors will be much harder to read.

What we can do is we'd also provide an identity function called `createDefinition` then users can use it like this -
```ts
let d = createDefintion(...) // more readable errors
useStateMachine(d)
```

**Pull request closed without merging**